### PR TITLE
Feature/jobfs

### PR DESCRIPTION
**Pull request by wiwichips**
Initial implementation of JobFS - to be re-written in JS in dcp-client

---

The API is

`JobFS`
- Constructor for Job Filesystems
- Takes no arguments

JobFS.InvalidPath (unspec'd)
- Exception class
- instances of this exception are raised when calling chdir with an invalid path

job_fs.add(src: (path | content), optional destination: (relative path | full path))
- Adds a file from local disc, directory from local disc, or a blob of data, onto the jobfs vfs at a specified location
- Can create a file anywhere, will create the necessary directories

job_fs.chdir(dest: path)
- Changes the current directory from which relative paths are resolved from to the specified path
- Raises an InvalidPath error when switching to a path that doesn't exist

_Everything else is just helpers_

### CODE REVIEW

**Comment by wiwichips** [comment]
NOTE: jobfs does not work with dcp until https://gitlab.com/Distributed-Compute-Protocol/dcp-client/-/merge_requests/222 is merged

**Comment by Severn-J-Lortie** [comment]
I think that jobfs.py should be renamed to job_fs.py for consistency

**Comment by wiwichips** [comment]
> I think that jobfs.py should be renamed to job_fs.py for consistency

@Severn-J-Lortie 
341a542ddc4aab0caad0e673f52a74659d9ae3a9

**Comment by wiwichips** [comment]
The API is

`JobFS`
- Constructor for Job Filesystems
- Takes no arguments

JobFS.InvalidPath
- Exception class
- instances of this exception are raised when calling chdir with an invalid path

job_fs.add(src: (path | content), optional destination: (relative path | full path))
- Adds a file from local disc, directory from local disc, or a blob of data, onto the jobfs vfs at a specified location
- Can create a file anywhere, will create the necessary directories

job_fs.chdir(dest: path)
- Changes the current directory from which relative paths are resolved from to the specified path
- Raises an InvalidPath error when switching to a path that doesn't exist

_Everything else is just helpers_

File: dcp/api/job_fs.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        """
+        if isinstance(local_src, io.IOBase):
+            local_src = local_src.read()
+
+        if isinstance(local_src, bytes) or isinstance(local_src, bytearray):
+            if vfs_dest is None:
+                raise Exception('Must specify a destination file path to write to')
+        elif isinstance(local_src, str) or isinstance(local_src, pathlib.PurePath):
+            if vfs_dest is None:
+                vfs_dest = os.path.join(self.home, os.path.basename(local_src))
```
Comment (Severn-J-Lortie) [inline dcp/api/job_fs.py:53]: Probably should be added from the current working directory, not home

File: dcp/api/job_fs.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    def __init__(self):
+        self.vfs = {}
+        self.vfs['/'] = self.vfs
+        self.home = '/home/pyodide'
+        self.mkdir(self.home)
+        self.cwd = self.home
+
+    def add(self, local_src, vfs_dest=None):
+        """Adds local file[s] to the vfs.
+        If dest not specified, add file from root.
```
Comment (Severn-J-Lortie) [inline dcp/api/job_fs.py:32]: Should be added from current working dir, not home

**Comment by Severn-J-Lortie** [comment]
LGTM!

File: dcp/api/job_fs.py (hunk 1 above)
Comment (wiwichips) [reply dcp/api/job_fs.py]: 0a46495e06ddee99f35a159b9569b36a04b02be2

File: dcp/api/job_fs.py (hunk 2 above)
Comment (wiwichips) [reply dcp/api/job_fs.py]: 0a46495e06ddee99f35a159b9569b36a04b02be2

**Pull request merged**